### PR TITLE
HOTT-868 - Rules of origin queries

### PR DESCRIPTION
**Pull request by jebw**
### Jira link

[HOTT-868](https://transformuk.atlassian.net/browse/HOTT-868)

### What?

I have added/removed/altered:

- [x] Added a `RulesOfOrigin::DataSet` model to group the various file based data sources together
- [x] Added a `RulesOfOrigin::Query` model to query the data set model
- [x] Fixed several edge case bugs in the other RoO data models
- [x] Added various factories to make it easier to generate data for testing

### Why?

I am doing this because:

- We need a way to query our file based data so that we can return the Rules of Origin data in the API

### CODE REVIEW

File: app/models/rules_of_origin/rule.rb [hunk 1]
```diff
@@ -3,16 +3,29 @@
...
   class Rule
     include ActiveModel::Model
+    include ActiveModel::Attributes
 
     ID_RULE_FORMAT = %r{\A\d+\z}.freeze
     SCHEME_CODE_FORMAT = %r{\A[a-zA-Z\-]+\z}.freeze
 
-    attr_accessor :id_rule, :scheme_code, :heading, :description,
-                  :quota_amount, :quota_unit, :rule, :alternate_rule
+    attribute :id_rule
```
Comment (willfish) [inline app/models/rules_of_origin/rule.rb:11]: :heart:

File: spec/factories/rules_of_origin/data_set_factory.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
+FactoryBot.define do
+  factory :rules_of_origin_data_set, class: 'RulesOfOrigin::DataSet' do
+    initialize_with { new scheme_set, rule_set, heading_mappings }
```
Comment (willfish) [inline spec/factories/rules_of_origin/data_set_factory.rb:3]: Crazy eval wizardry :dancers:

File: spec/models/rules_of_origin/heading_mappings_spec.rb [hunk 3]
```diff
@@ -61,13 +61,13 @@
     context 'with unknown scheme code' do
       let(:scheme_code) { 'unknown' }
 
-      it { is_expected.to be_empty }
+      it { is_expected.to eq({}) }
```
Comment (willfish) [inline spec/models/rules_of_origin/heading_mappings_spec.rb:64]: :+1:

Comment (jebw) [reply spec/models/rules_of_origin/heading_mappings_spec.rb:64]: Found I was returning an empty array - when results should be a hash - realised the weakness of checking for `empty?` but not type

**Pull request merged**